gitignore-parser.js

```diff
@@ -0,0 +1,39 @@
+'use strict';
+const path = require('path');
+const parseGitignore = require('parse-gitignore');
+
+class GitignoreParser
+{
+	constructor(opts) {
+		this.opts = Object.assign({
+			cwd: ''
+		}, opts);
+
+		this.parseFile = this.parseFile.bind(this);
+		this.parsePattern = this.parsePattern.bind(this);
+	}
+
+	parseFile(filepath) {
+		filepath = path.resolve(this.opts.cwd, filepath);
+		const baseDirectory = path.relative(this.opts.cwd, path.dirname(filepath));
+
+		return parseGitignore(filepath)
+			.map(pattern => this.parsePattern(pattern, baseDirectory))
+			.sort(patternInfo => patternInfo.isIgnorePattern ? 1 : -1)
+			.map(patternInfo => patternInfo.pattern);
+	}
+
+	parsePattern(pattern, baseDirectory) {
+		const isIgnorePattern = !pattern.startsWith('!');
+
+		if (isIgnorePattern) {
+			pattern = '!' + path.join(baseDirectory, pattern);
+		} else {
+			pattern = path.join(baseDirectory, pattern.substr(1));
+		}
+
+		return {isIgnorePattern, pattern};
```

Comment:
I've had the same idea, but that wouldn't work with double negated patterns (`!!foo.js`) – would it?

Hmm, the [gitignore man](https://git-scm.com/docs/gitignore) does not explain how these patterns are interpreted; hence I've given it a shot:

```sh
$ ls -A
foo.js  !foo.js  !!foo.js  !!!foo.js  .git  .gitignore

$ cat .gitignore 
*.js
!!foo.js

$ git status
On branch master

Initial commit

Untracked files:
  (use "git add <file>..." to include in what will be committed)

	!foo.js
	.gitignore

nothing added to commit but untracked files present (use "git add" to track)
```

It definitely needs a unit-test anyway.